docs/design/2020-06-27-non-recursive-common-table-expression.md

```diff
@@ -0,0 +1,85 @@
+
+# Proposal: Non-Recursive Common Table Expression
+
+- Author(s):     [Pingyu](https://github.com/pingyu) (Ping Yu)
+- Last updated:  2020-06-29
+- Discussion at: https://github.com/pingcap/tidb/issues/17472
+
+## Abstract
+
+This proposal proposes to support the __Non-recursive Common Table Expression (CTE)__.
+
+## Background
+
+From MySQL User Document, [13.2.15 WITH (Common Table Expressions)](https://dev.mysql.com/doc/refman/8.0/en/with.html):
+> A common table expression (CTE) is a named temporary result set that exists within the scope of a single statement and that can be referred to later within that statement, possibly multiple times. The following discussion describes how to write statements that use CTEs.
+
+E.g.
+```sql
+WITH
+  cte1 AS (SELECT a, b FROM table1),
+  cte2 AS (SELECT c, d FROM table2)
+SELECT b, d FROM cte1 JOIN cte2
+WHERE cte1.a = cte2.c;
+```
+
+## Proposal
+
+We design two strategies for CTE: __Inline__ & __Materialization__:
+
+#### Inline
+The CTE is expanded on where it's referenced:
+```sql
+WITH cte(k, v) AS (SELECT i, j FROM t1)
+SELECT v FROM cte WHERE k > 10
+UNION DISTINCT
+SELECT v FROM cte WHERE k < 100
+```
+After "inline":
+```sql
+SELECT v FROM (SELECT i AS k, j AS v FROM t1) WHERE k > 10
+UNION DISTINCT
+SELECT v FROM (SELECT i AS k, j AS v FROM t1) WHERE k < 100
+```
+
+We will add some rules to rewrite the plan trees, similar to `subquery optimization`[2].
+
+#### Materialization
+A CTE is executed first, and the result is stored temporarily during the statement execution.
+
+And on each point the CTE is referenced, the tuples of CTE is read from the temporary storage.
+
+To improve performance, the temporay result is stored in memory first, then spill to disk if out of memory quota, similar to disk-based executors[3].
+
+Besides, [4] introduces a `Producer-Consumer` model. We can utilize this model to make writing and reading of CTE parallel.
+
+![cte01](imgs/cte01.png)
+
+#### Optimization
+For `Inline` strategy, most existed rules for subquery can be reused.
+
+For `Materialization` strategy, disjunction of all predicates on top can be push down to CTE [4].
+
+Finally, the costs should be estimated, to determine which strategy is better for each CTE.
+
+## Rationale
+
+Mainstream database systems utilize `Inline` and `Materialization` to implement CTE, including MySQL[1], MariaDB[2] and Greenplum[4].
+
+## Implementation
+
+1. Stage 1: `Inline` should be not difficult. A rewrite rule is enough, similar to rewriting subquery[2].
+2. Stage 2: Implement `Materialization`, by storing chunks temporarily in memory or disk.
```

Comment:
Would the materialization phase slow down the entire execution of non-recursive CTE?

Reply:
`Materialization` can be involved in not only `CTE`, but also `View` and `Apply`. So I may suggest delaying implementing it and consider it as a standalone task.

Reply:
I think so. So I think materialization should be carefully optimized, e.g. predicates push-down, using memory if possible. @zz-jason

Reply:
agreed. we need a more detailed discussion about this optimization. I left some of my thoughts in the `Materialization` section.